x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/gen/script/Scripts.java

```diff
@@ -107,4 +113,136 @@ public static String classPackageAsPrefix(Class<?> function) {
+     */
+    public static ScriptTemplate multiValueDocValuesRewrite(ScriptTemplate script) {
+        return docValuesRewrite(script, false);
+    }
+
+    private static ScriptTemplate docValuesRewrite(ScriptTemplate script, boolean useSameValueInScript) {
+        int index = 0; // counts how many params there are to be able to get their value from script's params
+        Map<String, Object> params = script.params().asParams();
+        List<Object> fieldVars = new ArrayList<>();
+        List<Object> otherVars = new ArrayList<>();
+        StringBuilder newTemplate = new StringBuilder();
+        int negated = 0;
+
+        String[] tokens = splitWithMatches(script.template(), qlDocValuePattern);
+        for (String token : tokens) {
+            // A docValue call will be replaced with "X" followed by a counter
+            // The scripts that come into this method can contain multiple docValue calls for the same field
+            // This method will use only one variable for one docValue call
+            if ("InternalQlScriptUtils.docValue(doc,params.%s)".equals(token)) {
+                Object fieldName = params.get("v" + index);
+                
+                if (useSameValueInScript) {
+                    // if the field is already in our list, don't add it one more time
+                    if (fieldVars.contains(fieldName) == false) {
+                        fieldVars.add(fieldName);
+                    }
+                    newTemplate.append("X" + fieldVars.indexOf(fieldName));
+                } else {
+                    fieldVars.add(fieldName);
+                    newTemplate.append("X" + (fieldVars.size() - 1));
+                }
+                // increase the params position
+                index++;
+            } else if ("InternalQlScriptUtils.not(".equals(token)) {
+                negated++;
+            } else if ("params.%s".equals(token)) {
+                newTemplate.append(token);
+                // gather the other type of params (which are not docValues calls) so that at the end we rebuild the list of params
+                otherVars.add(params.get("v" + index));
+                index++;
+            } else {
+                newTemplate.append(token);
+            }
+            if (negated > 0) {
+                for (int i = 0; i < negated - 1; i++) {
+                    // remove this many closing parantheses as "InternalQlScriptUtils.not(" matches found minus one
+                    newTemplate.deleteCharAt(newTemplate.length() - 1);
+                }
+            }
+        }
+
+        // iterate over the fields in reverse order and add a multiValueDocValues call for each
+        for(int i = fieldVars.size() - 1; i >= 0; i--) {
+            newTemplate.insert(0, "InternalEqlScriptUtils.multiValueDocValues(doc,params.%s,X" + i + " -> ");
+            newTemplate.append(")");
+        }
+        if (negated > 0) {
```

Comment:
In the rest of this method, it looks like the translation supports multiple `not` in the script but here it assumes only one global negation.

Are there other cases that include multiple `not`? I couldn't come up with a query that would be translated into nested or multiple `not` and break the logic here though.

Could `negated` also be a boolean flag?

Reply:
Indeed there is one use case, rather obscure but possible at the moment, with `string` function that opens the door to infinite number of possibilities of `and`s an `or`s and `not`s: `where string(fieldA > 2 and not fieldB < 5 or not concat(host_name, port) == "abc:123") == "true"`.